ovos_PHAL_plugin_oauth/__init__.py

```diff
@@ -147,23 +166,105 @@ def oauth_callback(self, munged_id):
+            ).json()
 
         with OAuthTokenDatabase() as db:
             db.add_token(munged_id, token_response)
 
+        # Allow any registered app / skill to handle the token response urgently, if needed
+        # For example temporary tokens to generate a long lived token for the skill/plugin
+        app.bus.emit(
+            Message(f"oauth.token.response.{munged_id}", data=token_response))
+
         return params
 
+    def handle_get_app_host_info(self, message):
+        self.bus.emit(message.reply("oauth.app.host.info.response", {
+            "host": get_ip(),
+            "port": self.port
+        }))
+
+    #### QR Code Generation and Service Support ####
+    def handle_generate_qr(self, message):
+        skill_id = message.data.get("skill_id")
+        app_id = message.data.get("app_id")
+        munged_id = f"{skill_id}_{app_id}"  # key for oauth db
+        data = OAuthApplicationDatabase()[munged_id]
+        oauth_url = data.get("auth_endpoint")
+        client_id = data.get("client_id", None)
+        client_secret = data.get("client_secret", None)
+
+        if not oauth_url:
+            LOG.error(f"No auth endpoint found for oauth app {munged_id}")
+            return
+
+        plugin_service_url = self.build_plugin_service_url(
+            oauth_url, skill_id, app_id, client_id, client_secret)
+        self.bus.emit(message.reply("oauth.generate.qr.response", {
+            "skill_id": skill_id,
+            "app_id": app_id,
+            "qr": self.generate_qr(plugin_service_url, skill_id, app_id)
+        }))
+
+    def generate_qr(self, url, skill_id, app_id):
+        qr = qrcode.QRCode(
+            version=1,
+            error_correction=qrcode.constants.ERROR_CORRECT_L,
+            box_size=10,
+            border=4,
+        )
+        qr.add_data(url)
+        qr.make(fit=True)
+        img = qr.make_image(fill_color="black", back_color="white")
+        img_id = str(uuid.uuid4().hex)[:4]
+        temp_dir = tempfile.gettempdir()
+        img.save(f"{temp_dir}/{skill_id}_{app_id}_oauth_qr_{img_id}.png")
+        return f"{temp_dir}/{skill_id}_{app_id}_oauth_qr_{img_id}.png"
+
+    def build_plugin_service_url(self, oauth_url, skill_id, app_id, client_id=None, client_secret=None):
+        redirect_uri = f"http://{self.local_flask_host}:{self.port}/auth/callback/{skill_id}_{app_id}"
+        oauth_complete_url = f"{oauth_url}?redirect_uri={redirect_uri}"
+
+        if client_secret:
+            # Some services require client_id and client_secret
+            oauth_complete_url = f"{oauth_url}?client_id={client_id}&client_secret={client_secret}&redirect_uri={redirect_uri}"
+
+        if client_id:
+            # Some require only client_id
+            oauth_complete_url = f"{oauth_url}?client_id={client_id}&redirect_uri={redirect_uri}"
+
+        return oauth_complete_url
+
     def run(self):
-        app.run(port=self.port, debug=False)
+        self.plugin_started_event = threading.Event()
```

Comment:
this should be defined in init method

Reply:
if this is set in the init method it leads to the same issue in that it was having with getting the self.port 

```
  File "/home/aix/ovos/ovos-core/.venv/lib/python3.10/site-packages/ovos_PHAL_plugin_oauth/__init__.py", line 255, in run
    while not self.plugin_started_event.is_set():
AttributeError: 'OAuthPlugin' object has no attribute 'plugin_started_event'
```

Reply:
ha, i understand what you are trying to do now, this is the wrong approach then, the event should be completely removed

the issue is the call to super() being the very first line in init, you just need to move it to last after all things are defined and then they will be available in the run() method

Reply:
run() almost always executes before the __init__ method has even a chance to complete, so essentially self.plugin_started_event is not defined for run() on startup.

Reply:
thats because of the call to super(), just move it to bottom